dulwich/line_ending.py

```diff
@@ -229,19 +232,29 @@ def __init__(self, config_stack, gitattributes):
             core_eol, core_autocrlf, self.gitattributes
         )
 
-    def checkin_normalize(self, blob, tree_path):
+    def checkin_normalize(self, blob, tree_path, new_file):
```

Comment:
Couldn't we just call checkin_normalize for new files only?

Reply:
Places that calls `checkin_normalize` don't actually know that the callback is checkin_normalize. It was designed this way to support more filter callbacks in order to support Git custom filters in gitattributes.

Reply:
All filters should work at the same moments. See https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes

Reply:
I'm not sure if line-endings are exactly like filters in CGit but they operate at about the same time